sdk/python/kfp/components/graph_component.py

```diff
@@ -85,3 +89,165 @@ def pipeline_spec(self) -> pipeline_spec_pb2.PipelineSpec:
+                                    component_name, component_spec,
+                                    component_spec.executor_label,
+                                    original_components_executor_spec)
+
+            # Process the pipeline spec
+            for component in components_with_clones.keys():
+                corresponding_executor_name = components_with_clones[
+                    component].executor_name
+                del pipeline_spec.components[component]
+                del pipeline_spec.deployment_spec.fields[
+                    'executors'].struct_value.fields[
+                        corresponding_executor_name]
+
+            for component_name, clone_components in clone_mapping.items():
+                clones = clone_components + [component_name]
+
+                corresponding_executor_name, component_spec, executor_spec = components_with_clones[
+                    component_name].get_data()
+
+                pipeline_spec.components[component_name].CopyFrom(
+                    component_spec)
+                pipeline_spec.components[
+                    component_name].executor_label = corresponding_executor_name
+                pipeline_spec.deployment_spec.fields[
+                    'executors'].struct_value.fields[
+                        corresponding_executor_name].CopyFrom(executor_spec)
+
+                sorted_dag_task_spec = dict(
+                    sorted(pipeline_spec.root.dag.tasks.items()))
+                for task_spec in sorted_dag_task_spec.values():
+                    if task_spec.component_ref.name in clones:
+                        task_spec.component_ref.name = component_name
+
+                # for inner task group calling on components
+                sorted_components = dict(
+                    sorted(pipeline_spec.components.items()))
+                for component_spec in sorted_components.values():
+                    if component_spec.dag:
+                        for task_spec in component_spec.dag.tasks.values():
+                            if task_spec.component_ref.name in clones:
+                                task_spec.component_ref.name = component_name
+
+        def _clean_up_component_spec_names():
+            # clean up other component spec names
+            if not components_with_clones:
+                return
+            changed_names = {}
+            for component_name, component_spec in sorted(
+                    pipeline_spec.components.items()):
+                if component_name not in components_with_clones and component_spec.executor_label:
+                    last_delimiter = component_name.rfind('-')
+                    if len(
+                            component_name
+                    ) > last_delimiter + 1 and component_name[last_delimiter +
+                                                              1:].isnumeric():
+                        initial = component_name
```

Comment:
Consider extracting some of the subroutines into helper functions throughout. For example, this could be `get_name_from_component_ref` or something similar.